src/groups/bmq/bmqp/bmqp_putmessageiterator.cpp

```diff
@@ -581,13 +582,20 @@ int PutMessageIterator::next()
         return (rc * 10 + rc_INVALID_APPLICATION_DATA_OFFSET);  // RETURN
     }
 
-    bmqt::CompressionAlgorithmType::Enum cat =
-        d_header.compressionAlgorithmType();
-    int        flags = d_header.flags();
+    const int  flags = d_header.flags();
     const bool haveMPs =
         PutHeaderFlagUtil::isSet(flags, PutHeaderFlags::e_MESSAGE_PROPERTIES);
     const bool haveNewMPs = MessagePropertiesInfo::hasSchema(d_header);
-    int        length     = compressedApplicationDataSize();
+
+    // Validation: if new message properties schema is set, the flag
+    // e_MESSAGE_PROPERTIES must also be set.
+    if (BSLS_PERFORMANCEHINT_PREDICT_UNLIKELY(!haveMPs && haveNewMPs)) {
+        BSLS_PERFORMANCEHINT_UNLIKELY_HINT;
+        d_advanceLength = -1;
+        return rc_INVALID_MESSAGE_PROPERTIES;  // RETURN
+    }
```

Comment:
Fail if this is true: `!haveMPs && haveNewMPs`
Let's revert this expression and simplify it:
`!(!haveMPs && haveNewMPs)` ->
`(!!haveMPs) || (!haveNewMPs)` ->
`haveMPs || (!haveNewMPs)` ->
`haveMPs || (!(wire.value() != k_NO_WIRE_SCHEMA))` ->
`haveMPs || (wire.value() == k_NO_WIRE_SCHEMA)` ->
`haveMPs || wire.value() == k_NO_WIRE_SCHEMA`

So, the reverted expression here is equivalent to the offending failing assertion:
`BSLS_ASSERT_SAFE(d_isPresent || d_schemaWireId == k_NO_WIRE_SCHEMA);`